### PR TITLE
Fix PG search indexer connection exhaustion issue

### LINKED ISSUE #17278: 🐛 Bug Report: PgSearchEngineIndexer does not reliably release database connections when errors occur earlier in the pipeline
### 📜 Description

The `PgSearchEngineIndexer` uses one transaction for each indexing run. This transaction is committed when finalizing the index operation, and has various calls to `rollback` when errors occur within the indexer.

However, it appears that when _other parts_ of the indexing pipeline throw errors, none of the error handling code in the indexer is hit, and so the transaction is not committed or rolled back.

This means that over time, the search backend continues to accumulate database connections, until no more are available.

### 👍 Expected behavior

The indexer (and/or the indexing system itself) should be updated to guarantee that `finalize` (or another hook) is called whenever indexing stops for any reason, so that the connection can be freed. In particular, when other parts of the indexing pipeline throw errors, it should not cause the indexer to leak connections.

### 👎 Actual Behavior with Screenshots

When errors occur in (say) decorators, as long as the indexer has been initialized, the connection is not freed, and the number of connections held by the search backend increases until no more are available, at which time messages like this start to appear:

```
2023-04-06T12:09:05.663Z backstage warn Database keepalive failed for plugin search, KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
```

Note: to check the number of active connections, you can use the following query:

```SQL
SELECT * FROM pg_stat_activity WHERE datname='backstage_plugin_search';
```

### 👟 Reproduction steps

The [branch `search-module-pg-connection-repro`](https://github.com/backstage/backstage/compare/search-module-pg-connection-repro) contains a reproduction of the issue. See the commit messages in the branch for specific details of the changes made to reproduce.

1. Check out `search-module-pg-connection-repro`
2. Configure a postgres database
3. Start the backend

Once the backend is running, note the periodic errors logged to the console, and watch the number of connections increase by using the DB query above.

### 📃 Provide the context for the Bug.

_No response_

### 🖥️ Your Environment

_No response_

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, I don't have time to work on this right now

### PR DESCRIPTION
**Pull request by iamEAP**
## What / Why

Closes #17278

Note, I've opened https://github.com/backstage/backstage/issues/17291 for consideration on better ways to handle this.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Tests for new functionality and regression tests for bug fixes
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.test.ts [hunk 1]
```diff
@@ -156,4 +157,37 @@ describe('PgSearchEngineIndexer', () => {
     expect(result.error).toBe(expectedError);
     expect(tx.rollback).toHaveBeenCalledWith(expectedError);
   });
+
+  it('should rollback transaction on upstream error', async () => {
```
Comment (mtlewis) [inline plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.test.ts:161]: Nice work getting a test in for this! Could maybe include additional tests for the cases where rollback should _not_ be attempted? (i.e. when there's no error, and when rollback has already happened after the addition of `isCompleted` below).

File: plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.ts [hunk 2]
```diff
@@ -92,4 +92,32 @@ export class PgSearchEngineIndexer extends BatchSearchEngineIndexer {
...
+   */
+  async _destroy(error: Error | null, done: (error?: Error | null) => void) {
+    // Ignore situations where there was no error.
+    if (!error) {
+      done();
+      return;
+    }
+
+    try {
+      this.tx!.rollback(error);
```
Comment (mtlewis) [inline plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.ts:117]: We should probably `await` this, right?

Also, how about using `isCompleted` (TIL, cf. [knex transaction docs](https://knexjs.org/guide/transactions.html)):

```typescript
if (!this.tx!.isCompleted()) {
  await this.tx!.rollback();
}
```

**Review by mtlewis — COMMENTED** [review]
🚀 Looking great! Just a couple of small suggestions from me.

File: plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.ts (hunk 2 above)
Comment (iamEAP) [reply plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.ts:117]: Oooh, nice.  I wish the knex TS had better hints--it's unclear without looking at the docs that rollback returns a promise.

I notice all of the other rollback calls are also not awaited in this class.  Think it's worth adding that in elsewhere too?

File: plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.test.ts (hunk 1 above)
Comment (iamEAP) [reply plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.test.ts:161]: Makes sense!  I'll add a `.not.toHaveBeenCalled()` to the "happy path" case and add another case for the `isCompleted` scenario.

Comment (iamEAP) [reply plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.test.ts:170]: In addition to that happy path assertion, it ended up being simpler (if a little inelegant) to just add `expect(tx.rollback).toHaveBeenCalledTimes(1);` assertions to the existing error test cases.

Tests with the `TestPipeline` utility end up being more integration-y than unit-y, so I think it's fine.  I feel confident, at least!

File: plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.ts (hunk 2 above)
Comment (iamEAP) [reply plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.ts]: For the sake of focus/simplicity, I avoided touching other parts of the class and went with this suggestion verbatim.  If there are problems that arise where rollback itself fails, I'd hope we'd get report of that, and we could solve that all in one go.

**Comment by camilaibs** [comment]
Starting reviewing

File: plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.ts [hunk 3]
```diff
@@ -92,4 +92,31 @@ export class PgSearchEngineIndexer extends BatchSearchEngineIndexer {
...
+  /**
+   * Custom handler covering the case where an error occurred somewhere else in
+   * the indexing pipeline (e.g. a collator or decorator). In such cases, the
+   * finalize method is not called, which leaves a dangling transaction and
+   * therefore an open connection to PG. This handler ensures we close the
+   * transaction and associated connection.
+   *
+   * todo(@backstage/discoverability-maintainers): Consider introducing a more
+   * formal mechanism for handling such errors in BatchSearchEngineIndexer and
+   * replacing this method with it. See: #17291
```
Comment (camilaibs) [inline plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.ts:105]: Thank you for opening a issue for this to-do 🙌🏻

File: plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngine.ts [hunk 4]
```diff
@@ -285,6 +285,9 @@ export class ElasticSearchSearchEngine implements SearchEngine {
     });
 
     // Attempt cleanup upon failure.
+    // todo(@backstage/discoverability-maintainers): Consider introducing a more
```
Comment (camilaibs) [inline plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngine.ts:288]: Thank you for adding this to-do comment!

**Review by camilaibs — APPROVED** [review]
LGTM, thanks for helping on this @iamEAP!

**Pull request merged**